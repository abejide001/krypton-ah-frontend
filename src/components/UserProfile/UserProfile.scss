.ui.form.profileForm {
  .fields {
    margin: 0;
    margin-top: 1.2em;
    .field {
      padding-left: 0;
      padding-right: 0;
      margin-top: 1em;
    }
  }

  .field {
    label,
    .default.text,
    ::placeholder {
      text-transform: uppercase;
    }
  }

  .fields .input input,
  textarea,
  .dropdown {
    border-right: 0;
    border-left: 0;
    border-top: 0;
    border-radius: 0;
    padding-left: 0;
    font-size: 1.3rem;
    margin-top: -0.5rem;
    opacity: 1;
  }
  .field .input,
  .fields .field .dropdown.ui.selection {
    width: 90% !important;
  }

  .ui.dropdown > .text {
    font-weight: 400;
  }

  .btnRight {
    margin-top: 1rem;
  }
  div.bioField.field {
    margin-top: 2rem;
  }
  .bio {
    textarea {
      width: 100%;
      max-height: 4rem;
      height: 5rem;
    }
  }
  .field.btnRight {
    float: right;
    button {
      width: 18rem;
      height: 4rem;
      font-size: 1.4rem;
      margin-top: 1rem;
    }
  }
}

.form.ui.form.profileForm {
  .equal.width.fields {
    div.disabled.field {
      opacity: 1;
      div.disabled.input {
        opacity: 1;
        background-color: black;
        input[disabled]{
          opacity: 1;
        }
      }
    }
  }
  div.field.bioField {
    opacity: 1;
    div.disabled.field.bio{
      opacity: 1;
    }
  }
}
