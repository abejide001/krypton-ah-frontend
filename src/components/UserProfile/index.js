import React from 'react';
import PropTypes from 'prop-types';
import { Form } from 'semantic-ui-react';
import { profileFormGroups } from '../../constants/userProfile';
import './UserProfile.scss';

class UserProfile extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      firstname: '',
      lastname: '',
      country: '',
      phonenumber: '',
      gender: '',
      username: '',
      bio: ''
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleSelect = this.handleSelect.bind(this);
    this.renderFromGroups = this.renderFromGroups.bind(this);
    this.renderTextArea = this.renderTextArea.bind(this);
    this.renderUpdateButton = this.renderUpdateButton.bind(this);
  }

  componentWillMount() {
    const { profileData } = this.props;
    this.setState({ ...profileData });
  }

  handleChange(event) {
    const { name, value } = event.target;
    this.setState(state => ({
      ...state,
      [name.replace(' ', '')]: value
    }));
  }

  handleSelect(event, data) {
    const { name, value } = data;
    this.setState(state => ({
      ...state,
      [name]: value
    }));
  }

  handleSubmit(event) {
    event.preventDefault();
    const { updateUser } = this.props;
    updateUser(this.state);
  }

  renderFormElement(formElement, disabled) {
    const { element, name, type } = formElement;
    const { state } = this;
    return element === 'input' ? (
      <Form.Input
          disabled={disabled}
          id={name}
          size="massive"
          name={name}
          onChange={this.handleChange}
          placeholder={name.toUpperCase()}
          type={type}
          value={state[name.replace(' ', '')]}
        />
    ) : (
      <Form.Select
          disabled={disabled}
          id={name}
          name={name}
          value={state[name]}
          onChange={this.handleSelect}
          options={formElement.options}
          size="massive"
          placeholder={name.toUpperCase()}
        />
    );
  }

  renderFromGroups(disabled) {
    return profileFormGroups.map((formGroup) => {
      const [leftFormElement, rightFormElement] = formGroup;
      return (
        <Form.Group key={leftFormElement.name} widths="equal">
          <Form.Field>
            <label htmlFor={leftFormElement.name}>{leftFormElement.name.toLocaleUpperCase()}</label>
            {this.renderFormElement(leftFormElement, disabled)}
          </Form.Field>
          <Form.Field>
            <label htmlFor={rightFormElement.name}>
              {rightFormElement.name.toLocaleUpperCase()}
            </label>
            {this.renderFormElement(rightFormElement, disabled)}
          </Form.Field>
        </Form.Group>
      );
    });
  }

  renderTextArea(disabled, bio) {
    return (
      <Form.Field className="bioField">
        <label htmlFor="bio">BIO</label>
        <Form.TextArea
          id={bio}
          disabled={disabled}
          name="bio"
          onChange={this.handleChange}
          className="bio"
          autoHeight
          placeholder="Bio"
          value={bio}
        />
      </Form.Field>
    );
  }

  renderUpdateButton(user) {
    const updateButton = user === 'owner' ? (
      <Form.Button size="massive" className="btnRight" basic>
          UPDATE
      </Form.Button>
    ) : null;
    return updateButton;
  }

  render() {
    const { bio } = this.state;
    const { user } = this.props;
    let disabled = true,
      title = 'PROFILE';
    if (user === 'owner') {
      disabled = false;
      title = 'EDIT PROFILE';
    }
    return (
      <React.Fragment>
        <h3>{title}</h3>
        <Form onSubmit={this.handleSubmit} className="profileForm">
          {this.renderFromGroups(disabled)}
          {this.renderTextArea(disabled, bio)}
          {this.renderUpdateButton(user)}
        </Form>
      </React.Fragment>
    );
  }
}

UserProfile.defaultProps = {
  updateUser: () => {},
  user: ''
};

UserProfile.propTypes = {
  profileData: PropTypes.object.isRequired,
  updateUser: PropTypes.func,
  user: PropTypes.string
};

export default UserProfile;
