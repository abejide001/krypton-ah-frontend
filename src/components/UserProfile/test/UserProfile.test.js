import React from 'react';
import { shallow } from 'enzyme';
import UserProfile from '..';

describe('UserProfile', () => {
  it('should match snapshot', () => {
    const wrapper = shallow(<UserProfile profileData={{ firstname: 'Fistname' }} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('should match snapshot', () => {
    const wrapper = shallow(<UserProfile profileData={{ firstname: 'Fistname' }} user="owner" />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('UserProfile', () => {
  const event = {
      target: {
        name: 'firstname',
        value: 'Firstname',
      },
      preventDefault: () => {}
    },
    data = { name: 'gender', value: 'male' };
  describe('handleChange', () => {
    it('should set state property', () => {
      const wrapper = shallow(<UserProfile profileData={{ firstname: 'Fistname' }} />);
      wrapper.instance().handleChange(event);
      expect(wrapper.state().firstname).toEqual('Firstname');
    });
  });

  describe('handleSelect', () => {
    it('should set state property', () => {
      const wrapper = shallow(<UserProfile profileData={{ firstname: 'Fistname' }} />);
      wrapper.instance().handleSelect(event, data);
      expect(wrapper.state().gender).toEqual('male');
    });
  });

  describe('handleSelect', () => {
    it('should set state property', () => {
      const wrapper = shallow(<UserProfile profileData={{ firstname: 'Fistname' }} />);
      wrapper.instance().handleSelect(event, data);
      expect(wrapper.state().gender).toEqual('male');
    });
  });

  describe('handleSubmit', () => {
    it('should call updateUser', () => {
      const updateUserMock = jest.fn();
      const wrapper = shallow(<UserProfile updateUser={updateUserMock} profileData={{ firstname: 'Fistname' }} />);
      wrapper.instance().handleSubmit(event);
      expect(updateUserMock).toHaveBeenCalledTimes(1);
    });
  });
});
