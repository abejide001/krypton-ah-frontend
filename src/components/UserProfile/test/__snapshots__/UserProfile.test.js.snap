// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfile should match snapshot 1`] = `
<Fragment>
  <h3>
    PROFILE
  </h3>
  <Form
    as="form"
    className="profileForm"
    onSubmit={[Function]}
  >
    <FormGroup
      key="firstname"
      widths="equal"
    >
      <FormField>
        <label
          htmlFor="firstname"
        >
          firstname
        </label>
        <FormInput
          as={[Function]}
          control={[Function]}
          disabled={true}
          id="firstname"
          name="firstname"
          onChange={[Function]}
          placeholder="firstname"
          size="massive"
          type="text"
          value="Fistname"
        />
      </FormField>
      <FormField>
        <label
          htmlFor="lastname"
        >
          lastname
        </label>
        <FormInput
          as={[Function]}
          control={[Function]}
          disabled={true}
          id="lastname"
          name="lastname"
          onChange={[Function]}
          placeholder="lastname"
          size="massive"
          type="text"
          value=""
        />
      </FormField>
    </FormGroup>
    <FormGroup
      key="username"
      widths="equal"
    >
      <FormField>
        <label
          htmlFor="username"
        >
          username
        </label>
        <FormInput
          as={[Function]}
          control={[Function]}
          disabled={true}
          id="username"
          name="username"
          onChange={[Function]}
          placeholder="username"
          size="massive"
          type="text"
          value=""
        />
      </FormField>
      <FormField>
        <label
          htmlFor="gender"
        >
          gender
        </label>
        <FormSelect
          as={[Function]}
          control={[Function]}
          disabled={true}
          id="gender"
          name="gender"
          onChange={[Function]}
          options={
            Array [
              Object {
                "text": "male",
                "value": "male",
              },
              Object {
                "text": "female",
                "value": "female",
              },
            ]
          }
          placeholder="gender"
          size="massive"
          value=""
        />
      </FormField>
    </FormGroup>
    <FormGroup
      key="country"
      widths="equal"
    >
      <FormField>
        <label
          htmlFor="country"
        >
          country
        </label>
        <FormSelect
          as={[Function]}
          control={[Function]}
          disabled={true}
          id="country"
          name="country"
          onChange={[Function]}
          options={
            Array [
              Object {
                "text": "Nigeria",
                "value": "Nigeria",
              },
              Object {
                "text": "Ghana",
                "value": "Ghana",
              },
            ]
          }
          placeholder="country"
          size="massive"
          value=""
        />
      </FormField>
      <FormField>
        <label
          htmlFor="phone number"
        >
          phone number
        </label>
        <FormInput
          as={[Function]}
          control={[Function]}
          disabled={true}
          id="phone number"
          name="phone number"
          onChange={[Function]}
          placeholder="phone number"
          size="massive"
          type="tel"
          value=""
        />
      </FormField>
    </FormGroup>
    <FormField
      className="bioField"
    >
      <label
        htmlFor="bio"
      >
        BIO
      </label>
      <FormTextArea
        as={[Function]}
        autoHeight={true}
        className="bio"
        control={[Function]}
        disabled={true}
        id=""
        name="bio"
        onChange={[Function]}
        placeholder="Bio"
        value=""
      />
    </FormField>
  </Form>
</Fragment>
`;

exports[`UserProfile should match snapshot 2`] = `
<Fragment>
  <h3>
    EDIT PROFILE
  </h3>
  <Form
    as="form"
    className="profileForm"
    onSubmit={[Function]}
  >
    <FormGroup
      key="firstname"
      widths="equal"
    >
      <FormField>
        <label
          htmlFor="firstname"
        >
          firstname
        </label>
        <FormInput
          as={[Function]}
          control={[Function]}
          disabled={false}
          id="firstname"
          name="firstname"
          onChange={[Function]}
          placeholder="firstname"
          size="massive"
          type="text"
          value="Fistname"
        />
      </FormField>
      <FormField>
        <label
          htmlFor="lastname"
        >
          lastname
        </label>
        <FormInput
          as={[Function]}
          control={[Function]}
          disabled={false}
          id="lastname"
          name="lastname"
          onChange={[Function]}
          placeholder="lastname"
          size="massive"
          type="text"
          value=""
        />
      </FormField>
    </FormGroup>
    <FormGroup
      key="username"
      widths="equal"
    >
      <FormField>
        <label
          htmlFor="username"
        >
          username
        </label>
        <FormInput
          as={[Function]}
          control={[Function]}
          disabled={false}
          id="username"
          name="username"
          onChange={[Function]}
          placeholder="username"
          size="massive"
          type="text"
          value=""
        />
      </FormField>
      <FormField>
        <label
          htmlFor="gender"
        >
          gender
        </label>
        <FormSelect
          as={[Function]}
          control={[Function]}
          disabled={false}
          id="gender"
          name="gender"
          onChange={[Function]}
          options={
            Array [
              Object {
                "text": "male",
                "value": "male",
              },
              Object {
                "text": "female",
                "value": "female",
              },
            ]
          }
          placeholder="gender"
          size="massive"
          value=""
        />
      </FormField>
    </FormGroup>
    <FormGroup
      key="country"
      widths="equal"
    >
      <FormField>
        <label
          htmlFor="country"
        >
          country
        </label>
        <FormSelect
          as={[Function]}
          control={[Function]}
          disabled={false}
          id="country"
          name="country"
          onChange={[Function]}
          options={
            Array [
              Object {
                "text": "Nigeria",
                "value": "Nigeria",
              },
              Object {
                "text": "Ghana",
                "value": "Ghana",
              },
            ]
          }
          placeholder="country"
          size="massive"
          value=""
        />
      </FormField>
      <FormField>
        <label
          htmlFor="phone number"
        >
          phone number
        </label>
        <FormInput
          as={[Function]}
          control={[Function]}
          disabled={false}
          id="phone number"
          name="phone number"
          onChange={[Function]}
          placeholder="phone number"
          size="massive"
          type="tel"
          value=""
        />
      </FormField>
    </FormGroup>
    <FormField
      className="bioField"
    >
      <label
        htmlFor="bio"
      >
        BIO
      </label>
      <FormTextArea
        as={[Function]}
        autoHeight={true}
        className="bio"
        control={[Function]}
        disabled={false}
        id=""
        name="bio"
        onChange={[Function]}
        placeholder="Bio"
        value=""
      />
    </FormField>
    <FormButton
      as={[Function]}
      basic={true}
      className="btnRight"
      control={[Function]}
      size="massive"
    >
      UPDATE
    </FormButton>
  </Form>
</Fragment>
`;
